### ISSUE TITLE
Purple annotation is wrongly regarded as Blue

### ISSUE THREAD

**Issue by vanbang9710** [mgmeyers/obsidian-zotero-integration#168]
I highlight with blue and purple
![image](https://user-images.githubusercontent.com/52906188/213923802-005a43dd-20c9-4245-bc13-4ea51e5f7ae3.png)
The color properties are still right, but colorCategory are both Blue
![image](https://user-images.githubusercontent.com/52906188/213923775-9cb6b207-e2e2-411a-a11c-283625194a81.png)

**Comment by mgmeyers** [comment]
@vanbang9710 Can you share the `color` of your blue and purple? Looks like `#2ea8e5` is the purple one. What's the blue?

**Comment by elorie** [comment]
#a28ea5

**Comment by Xervoo** [comment]
The issue is the "colorCategory" for purple annotations being "Blue". Note that the actual "color" key has two different values, matching the ones used by Zotero. Given that, I always assumed this was intended behaviour, since blue and purple are quite similar colour wise anyway. I'd recommend styling the document using that key instead for now.

**Comment by medium-old** [comment]
Hi @Xervoo , would you mind briefly explaining in layman terms? For example, is this something to be addressed in the css? if so how are you suggesting to differentiate the values? (non-code specialist asking)

**Comment by elorie** [comment]
> Hi @Xervoo , would you mind briefly explaining in layman terms? For example, is this something to be addressed in the css? if so how are you suggesting to differentiate the values? (non-code specialist asking)

If you want to do something foth those annotation colors in your note template in obsidian,  e.g. give them different background colors, you have to use the color as a hexvalue instead of the more human readable colorcategory to do different things with blue and purple annotations

**Comment by medium-old** [comment]
> > Hi @Xervoo , would you mind briefly explaining in layman terms? For example, is this something to be addressed in the css? if so how are you suggesting to differentiate the values? (non-code specialist asking)
> 
> If you want to do something with those annotation colors in your note template in obsidian, e.g. give them different background colors, you have to use the color as a hexvalue instead of the more human readable colorcategory to do different things with blue and purple annotations


Hi there! Thank you so much for explaining this. Unfortunately, any attempt to address these colors via hexvalue/alternative reference is not possible. The importation keeps reading Purple annotations as Blue (likely because the colorCategory is blue for both?), thus anything I try on the Obsidian end isn't responding.  

On a similar note - there are no issues with the 3 additional colors (magenta, orange, gray)!

**Comment by Xervoo** [comment]
Hi,
Since you mention other colours, I am led to believe that you don't use Zotero's built-in PDF editor? In that case, it will be much more difficult to help you, since the colours wouldn't be the standard ones used.
<img width="553" alt="image" src="https://user-images.githubusercontent.com/46051794/220930648-b5b67d45-4f1c-4c1f-ae5e-bcbdd8bfe1fa.png">
When using Zotero's built-in PDF editor, you can use the "color" key from your image above in the templates created.
An example:
`{% for a in annotations %}{#PURPLE}{% if a.color == "#a28ae5" %}{% set aColor = "purple" %}{#BLUE}{% elseif a.color == "#2ea8e5"%}{% set aColor = "blue" %}{% endif %}`
You would do this for all the colours that you highlight with / that show up in the data explorer when looking at the export from your document. Then you can use the "aColor" for whatever you want. The key is the if statement checking for the hex code in the color field of the annotation.

**Comment by elorie** [comment]
> @vanbang9710 Can you share the `color` of your blue and purple? Looks like `#2ea8e5` is the purple one. What's the blue?
One addition after digging a little bit in the source code: the h of the hsl value of purple is 256, the blue h is 200, changing the number in line 115 to 250 or 210 might fix it.

**Comment by medium-old** [comment]
@Xervoo Hi there! Thank you for explaining this further - your suggestion has clarified the matter but unfortunately I still can't manipulate Purple. Calling it quits :)  also, I am using Zotero Beta, which has released the aforementioned colors only a few days ago - check it out~

**Comment by Xervoo** [comment]
That's rather odd. I just got the update as well the other day and haven't tried to import any texts since. I gave it a shot now though and it still seems to work. I'll attach my template if you want to give that a shot and see if it works (you'll probably want to remove the "includes" before the citekey line.

[Import Template.md](https://github.com/mgmeyers/obsidian-zotero-integration/files/10838461/Import.Template.md)

**Comment by AllisonDai** [comment]
> @Xervoo Hi there! Thank you for explaining this further - your suggestion has clarified the matter but unfortunately I still can't manipulate Purple. Calling it quits :) also, I am using Zotero Beta, which has released the aforementioned colors only a few days ago - check it out~

I solved it! Here are my codes for the note section: 
`{% for annotation in annotations -%}{%- if annotation.annotatedText -%}{% if 'Red' in annotation.colorCategory %} 
##### {{annotation.annotatedText | escape }} ([{{annotation.page}}](zotero://open-pdf/library/items/{{annotation.attachment.itemKey}}?page={{annotation.page}}&annotation={{annotation.id}})) {% else%}
{% if '#a28ae5' in annotation.color %}
<mark class="customZot-Purple">{{annotation.annotatedText | escape }}</mark> ([{{annotation.page}}](zotero://open-pdf/library/items/{{annotation.attachment.itemKey}}?page={{annotation.page}}&annotation={{annotation.id}})) } {% else %}
<mark class="customZot-{% if annotation.color %}{{annotation.colorCategory}} {% endif %}">{{annotation.annotatedText | escape }}</mark> ([{{annotation.page}}](zotero://open-pdf/library/items/{{annotation.attachment.itemKey}}?page={{annotation.page}}&annotation={{annotation.id}}))

{% endif %}{%- endif %}{%- endif %} 
{% if annotation.imageRelativePath %} ![[{{annotation.imageRelativePath}}]]{% endif %}{% if annotation.comment %} 
>****Comment: {{annotation.comment}}***
{% endif %}{% endfor -%}
`
Here is my CSS: 
`.customZot-Yellow {
    background-color: rgba(255, 204, 0, 0.2) !important;
}

.customZot-Green {
    background-color: rgba(156, 175, 136, 0.2) !important;
}

.customZot-Blue {
    background-color: rgba(137, 207, 240, 0.2) !important;
}

.customZot-Purple {
    background-color: rgba(0, 0, 225, 0.2) !important;
}`

**Comment by mgmeyers** [comment]
Hey all, there really isn't a good fix for this, unfortunately. The issue is that where blue stops and purple begins on the color spectrum is subjective. If you need precise differentiation of colors, I recommend using the exact color values that show up in the data explorer rather than color categories.

**Comment by neouen** [comment]
I solved this problem by changing the default hex of purple color form“a28ae5”to“a73ae6” through advanced configuration of zotero，then the purple color can be correctly recognized.

**Comment by Sriramgireesh** [comment]
> I solved this problem by changing the default hex of purple color form“a28ae5”to“a73ae6” through advanced configuration of zotero，then the purple color can be correctly recognized.

Hi, I am sorta new to zotero. Could you kindly guide me with the steps to achieve this? Thanks a ton.

**Comment by egoipse** [comment]
Here's another solution: to call the "color" tag instead of "colorCategory" and create css for colors hex codes instead of color names.

In template, something like this:

```
{% for annotation in annotations %}
{% if annotation.comment %}
<mark class="hltr-{{annotation.color | lower | replace("#", "")}}">{{annotation.annotatedText}}</mark>{% endif %}{% endfor %}
``` 
In css classes' names, color name should be replaced by color hex codes (I use highlightr plugin for this):

``` 
.hltr-f19837 { /* orange */
	background-color: #f19837;
}

.hltr-ffd400 { /* yellow */
	background-color: #ffd400;
}

.hltr-5fb236 { /* green */
	background-color: #5fb236;
}

/* and so on for the rest of the colors */
```

**Comment by sulth1** [comment]
> Here's another solution: to call the "color" tag instead of "colorCategory" and create css for colors hex codes instead of color names.
> 
> In template, something like this:
> 
> ```
> {% for annotation in annotations %}
> {% if annotation.comment %}
> <mark class="hltr-{{annotation.color | lower | replace("#", "")}}">{{annotation.annotatedText}}</mark>{% endif %}{% endfor %}
> ```
> 
> In css classes' names, color name should be replaced by color hex codes (I use highlightr plugin for this):
> 
> ```
> .hltr-f19837 { /* orange */
> 	background-color: #f19837;
> }
> 
> .hltr-ffd400 { /* yellow */
> 	background-color: #ffd400;
> }
> 
> .hltr-5fb236 { /* green */
> 	background-color: #5fb236;
> }
> 
> /* and so on for the rest of the colors */
> ```

Hi. I don't understand the second part. I also use the highlightr plugin. I have tried to call the color tag instead of color category, but now all my annotations are bright yellow. I guess I miss something?

I have tried to switch "Yellow" with "#FFF3A3A6" in highlightr plugin, but it doesn't seem to be correct?

**Comment by jluckyiv** [comment]
@sulth1 the second code snippet is custom [CSS classes](https://developer.mozilla.org/en-US/docs/Web/CSS/Class_selectors). I haven’t seen your template, so I’m making some assumptions with my explanation: that you are unfamiliar with Highlightr’s inline CSS vs CSS classes approach, you are unfamiliar with custom CSS classes, and you are using a custom theme. The [Highlightr-Plugin README](https://github.com/chetachiezikeuzor/Highlightr-Plugin/blob/master/README.md) explains:

> Version [1.1.2](https://github.com/chetachiezikeuzor/Highlightr-Plugin/blob/master) adds the ability to choose between inline CSS and CSS classes. This will create a new stylesheet that will how all of your highlight colors. Each class is named hltr-${highlight_title_here} and are generated from the titles by which you have named your highlights. Although inline CSS is highly encouraged, CSS classes will make your highlights much more flexible and easier to customize.

I have noticed that all the questions and answers above seem to use CSS classes. I use inline CSS, so I’ll show my approach once I try to answer your question. (Some of this is guesswork because I don’t do it this way, but I might change my approach based on what I’ve read.) I believe you can copy and paste @egoipse ‘s custom CSS into a [CSS snippet](https://help.obsidian.md/Extending+Obsidian/CSS+snippets). CSS snippets are stored in `{vault-folder}/.obsidian/snippets/`, so you could save yours in `{vault-folder}/.obsidian/snippets/zotero-highlight.css`.

Another way to use custom CSS classes is to modify your current theme. Find your theme’s CSS file in `{vault-folder}/.obsidian/themes/{custom-theme-name}/theme.css` file. For instance (on Mac/Linux), mine is `~/Vault/.obsidian/themes/Tokyo Night/theme.css`. I suggest copying the entire theme folder (e.g., `Tokyo Night Custom`), switching to the theme copy in Obsidian, and modifying the copy. The downside of this approach is that you will have to redo this customization if you change themes, so the snippet approach might be better. 

My approach is inline CSS, which I put directly in the template. First, create a JavaScript object in your template above your highlights section. The keys on the left are the Zotero default highlight values. The values on the right are taken from the Highlightr settings.

```jinja
<%*
let hl = {
  "#aaaaaa": "#C5C8D3A6",
  "#ff6666": "#FF5582A6",
  "#f19837": "#FFB86CA6",
  "#ffd400": "#FFF3A3A6",
  "#5fb236": "#BBFABBA6",
  "#2ea8e5": "#ADCCFFA6",
  "#a28ae5": "#D2B3FFA6",
  "#e56eee": "#FF91FCA6",
}
-%>
```

Second, use the values in your highlights section of the template. My snippet is below. The `hl[‘{{ annotation.color | lower }}’]` mapping converts the Zotero hex color to the Highlightr hex color.

```jinja
{% for annotation in annotations %}
{%- if annotation.annotatedText -%}
- <mark style="background: <% hl['{{ annotation.color | lower }}'] %>"> {{annotation.annotatedText}}</mark>
{%- endif %}
{% endfor %}
```

The advantage of the inline CSS approach is that it can be simpler. The disadvantage is that you have to duplicate this code in every template. Hope this answer helps.
